### PR TITLE
Plugins: add Crop effect

### PR DESCRIPTION
**Pull request by iam-TJ**
Crop uses slider controls to select left, top, width and height of area
to crop from the source frame. It maintains the same aspect ratio and
automatically reduces width and height when left or top positions are
increased. It sets a minimum width and height at 10% of the source.

This is the 'debug' version for initial sharing and testing. Once it is OKed the debug code will be removed.

Please test.

This addresses one part of issue #72.

### CODE REVIEW

**Comment by hipersayanX** [comment]
You are complicating the thing, the plugins does not need to depend neither of other plugins or keep the aspect ratio. Just create a bool property for editing, when it's true show the full picture and the marker, when false just show the cropped picture. then you add the properties: **top**, **bottom**, **left**, **right**, which define the limits of the rectangle. If **top < bottom** or **left > right**, just swap the values internally. The sliders and spinbox goes in the range [0, width - 1] ([0%, 100%]) and  [0, height - 1] ([0%, 100%]).

Crop design

File: libAvKys/Plugins/Crop/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+set(QT_COMPONENTS
+    Gui
+    Qml)
+find_package(QT NAMES Qt${QT_VERSION_MAJOR} COMPONENTS
+             ${QT_COMPONENTS}
+             REQUIRED)
+find_package(Qt${QT_VERSION_MAJOR} ${QT_MINIMUM_VERSION} COMPONENTS
+             ${QT_COMPONENTS}
+             REQUIRED)
+		 add_library(Crop SHARED
```
Comment (hipersayanX) [inline libAvKys/Plugins/Crop/CMakeLists.txt:39]: I have seen this before in other commits, and I have passed them silently, but please, keep the code clean, don't add unnecessary spaces in the code, it make hard to read and understand the code.
There is no reason for adding spaces before ```add_library(```.

File: libAvKys/Plugins/Crop/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+                          RUNTIME_OUTPUT_DIRECTORY ${CMAKE_BINARY_DIR}/build/${PLUGINSDIR})
+elseif (NOT ANDROID)
+    set_target_properties(Crop PROPERTIES
+                          LIBRARY_OUTPUT_DIRECTORY ${CMAKE_BINARY_DIR}/build/${PLUGINSDIR})
+endif ()
+
+add_dependencies(Crop avkys)
+target_include_directories(Crop
+                           PRIVATE ../../Lib/src)
+					   target_compile_definitions(Crop PRIVATE AVKYS_PLUGIN_CROP)
```
Comment (hipersayanX) [inline libAvKys/Plugins/Crop/CMakeLists.txt:58]: Same here.

File: libAvKys/Plugins/Crop/share/qml/main.qml [hunk 3]
```diff
@@ -0,0 +1,328 @@
...
+import QtQuick.Layouts 1.3
+import CropElement 1.0
+
+GridLayout {
+    property real aspectRatio: 0
+    property bool firstMax: true
+    property bool firstMin: true
+    property int tipDelay: 1000
+    property int tipTimeout: 5000
+    property rect source: Qt.rect(0,0,0,0)
```
Comment (hipersayanX) [inline libAvKys/Plugins/Crop/share/qml/main.qml:31]: https://github.com/webcamoid/webcamoid/wiki/Coding-style-and-conventions#functions-and-methods

File: libAvKys/Plugins/Crop/src/cropelement.h [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+               READ box
+               WRITE setBox
+               RESET resetBox
+               NOTIFY boxChanged)
+    Q_PROPERTY(QRectF minimum
+               READ minimum
+               WRITE setMinimum
+               RESET resetMinimum
+               NOTIFY minimumChanged)
+    // represents size of original frame; not changable externally
```
Comment (hipersayanX) [inline libAvKys/Plugins/Crop/src/cropelement.h:43]: If the properties are not supposed to be used externally, don't make them public, private things must be moved to the private class. Keep public just the minimal code to use the class.

File: libAvKys/Plugins/Crop/src/cropelement.h [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+    private:
+        CropElementPrivate *d;
+
+    protected:
+        enum Changed {
+            eLeft, eTop, eWidth, eHeight
+        };
+        QString controlInterfaceProvide(const QString &controlId) const;
+        void controlInterfaceConfigure(QQmlContext *context, const QString &controlId) const;
+        bool aspectRatioGood(QRect rect);
```
Comment (hipersayanX) [inline libAvKys/Plugins/Crop/src/cropelement.h:70]: Same here, move the code to the private class.

File: libAvKys/Plugins/Crop/src/cropelement.cpp [hunk 6]
```diff
@@ -0,0 +1,258 @@
...
+{
+    bool updateWidth = true;
+    bool updateHeight = true;
+    bool updateMax = false;
+
+    QRect newBox = qfBox.toRect();
+    if (newBox.isNull() || ! newBox.isValid() || newBox == d->m_box)
+        return;
+
+    int maxX = newBox.width() + newBox.left();
```
Comment (hipersayanX) [inline libAvKys/Plugins/Crop/src/cropelement.cpp:112]: https://github.com/webcamoid/webcamoid/wiki/Coding-style-and-conventions#blocks-of-code-loops-and-conditionals

File: libAvKys/Plugins/Crop/CMakeLists.txt (hunk 1 above)
Comment (iam-TJ) [reply libAvKys/Plugins/Crop/CMakeLists.txt:39]: That looks like some kind of spaces vs tabs glitch - its certainly not intentional. Not sure quite how those are getting in. I'll investigate QTCreator which I used for that, as opposed to `vim` for others.

File: libAvKys/Plugins/Crop/src/cropelement.h (hunk 4 above)
Comment (iam-TJ) [reply libAvKys/Plugins/Crop/src/cropelement.h:43]: This is deliberate in order to allow additional control from the GUI if requested.

**Comment by iam-TJ** [comment]
> You are complicating the thing, the plugins does not need to depend neither of other plugins or keep the aspect ratio. Just create a bool property for editing, when it's true show the full picture and the marker, when false just show the cropped picture. then you add the properties: **top**, **bottom**, **left**, **right**, which define the limits of the rectangle. If **top < bottom** or **left > right**, just swap the values internally. The sliders and spinbox goes in the range [0, width - 1] ([0%, 100%]) and [0, height - 1] ([0%, 100%]).

Free-for-all isn't the use-case I'm addressing. The use-case is "Jane wants a 800 pixel width frame that maintains ratios" or "Jane can select the starting point and the result will maintain ratios".  I'd argue that presenting Jane with width/height sliders is much more user friendly than expecting her to figure out the correct left and right, top and bottom, values to get her desired outcome.

I did it this way deliberately to avoid problems sending a modified video stream with improbable dimensions and aspect ratio since some arbitrary limits would then have to be placed on the acceptable values e.g. would 1920x1 be acceptable?  12x1080? 

How is this edit mode with rubber-band bounding rectangle initiated? I didn't find anything in `libAvKys/Lib/` for that.

File: libAvKys/Plugins/Crop/src/cropelement.h (hunk 4 above)
Comment (iam-TJ) [reply libAvKys/Plugins/Crop/src/cropelement.h:43]: I should add that both `minimum` and `maximum` are required for `NOTIFY` and, if the QML functionality is modified, `READ`.

**Comment by iam-TJ** [comment]
I think I see where the confusion is arising over functionality - I should have called this plugin "Letterbox" not "Crop. I'll redo it as "LetterBox". see https://en.wikipedia.org/wiki/Letterboxing_(filming)

**Comment by hipersayanX** [comment]
>Free-for-all isn't the use-case I'm addressing. The use-case is "Jane wants a 800 pixel width frame that maintains ratios" or "Jane can select the starting point and the result will maintain ratios". I'd argue that presenting Jane with width/height sliders is much more user friendly than expecting her to figure out the correct left and right, top and bottom, values to get her desired outcome.

But why limiting it, it make no sense, if you want keep the original aspect ratio of the image what you want to do is zoom in, not cropping. Cropping is supposed to cut the image ignoring the it's original aspect ratio, using the boundaries as parameters.

>I did it this way deliberately to avoid problems sending a modified video stream with improbable dimensions and aspect ratio since some arbitrary limits would then have to be placed on the acceptable values e.g. would 1920x1 be acceptable? 12x1080?

If the user want an image output of 1920x1 or 12x1080, what so bad about it?
Generalize the problem, then handle special cases.

>How is this edit mode with rubber-band bounding rectangle initiated? I didn't find anything in libAvKys/Lib/ for that.

Using a [QPainter::drawRect()](https://doc.qt.io/qt-5/qpainter.html#drawRect-1).

>I think I see where the confusion is arising over functionality - I should have called this plugin "Letterbox" not "Crop". I'll redo it as "LetterBox". see https://en.wikipedia.org/wiki/Letterboxing_(filming)

Ok, then what you want to do can be done with the [Cinema](https://github.com/webcamoid/webcamoid/tree/master/libAvKys/Plugins/Cinema) plugin, adding support for horizontal cropping then handling the special case into a Letterbox plugin.

**Comment by hipersayanX** [comment]
At the end I implemented the effect (944021b) as I said in [my last comment](https://github.com/webcamoid/webcamoid/pull/547#issuecomment-1160671854):

**Pull request closed without merging**